batchflow/tests/datasetindex_test.py

```diff
@@ -1,221 +1,294 @@
+    assert valid_share == 1
+
+def test_baseset_calc_split_correctness_4(all_indices):
+    dsi = all_indices()
+    length = len(dsi)
+    left = dsi.calc_split(shares=[0.5, 0.5])
+    right = (0.5*length, 0.5*length, 0)
+    assert left == right
+
+def test_build_index(all_indices):
+    """ True contents of 'dsi.index' are recovered from the constructor. """
+    dsi, struct = all_indices(struct=True)
+    if isinstance(struct, int):
+        struct = np.arange(struct)
+    elif isinstance(struct, DatasetIndex):
+        struct = struct.index
+    elif callable(struct):
+        struct = struct()
+    assert (dsi.index == struct).all()
+
+def test_shuffle_bool_false(all_indices):
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=False)
+    right = np.arange(len(dsi))
+    assert (left == right).all()
+
+def test_shuffle_bool_true(all_indices):
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=True)
+    right = np.arange(len(dsi))
+    assert (left != right).any()
+    assert set(left) == set(right)
+
+def test_shuffle_bool_int(all_indices):
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=SIZE)
+    right = np.arange(len(dsi))
+    assert (left != right).any()
+    assert set(left) == set(right)
+
+def test_shuffle_bool_randomstate(all_indices):
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=np.random.RandomState(SIZE))
+    right = np.arange(len(dsi))
+    assert (left != right).any()
+    assert set(left) == set(right)
+
+def test_shuffle_bool_cross(all_indices):
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=np.random.RandomState(SIZE))
+    right = dsi._shuffle(shuffle=SIZE)
+    assert (left == right).all()
+
+def test_shuffle_bool_callable(all_indices):
+    """ Callable 'shuffle' should return order. """
+    dsi = all_indices()
+    left = dsi._shuffle(shuffle=(lambda _: np.arange(len(dsi))))
+    right = np.arange(len(dsi))
+    assert (left == right).all()
+
+@pytest.mark.parametrize("repeat_time", [1]*1)
+def test_split_correctness(repeat_time, all_indices):
+    """ Constants in 'shares' are such that test does not raise errors. """
+    dsi = all_indices()
+    shares = .3 - np.random.random(3) *.05 *repeat_time
+    dsi.split(shares=shares)
+    assert len(dsi) == (len(dsi.train)
+                        + len(dsi.test)
+                        + len(dsi.validation))
+
+def test_create_batch_pos_false(all_indices):
+    """ When 'pos' is False, method returns the same, as its first argument. """
+    dsi = all_indices()
+    length = len(dsi)
+    left = dsi.create_batch(range(length), pos=False).index
+    right = range(length)
+    assert (left == right).all()
+
+def test_create_batch_type(all_indices):
+    """ Method 'create_batch' must be type-preserving. """
+    dsi = all_indices()
+    length = len(dsi)
+    assert isinstance(dsi.create_batch(range(length)), DatasetIndex)
+
+def test_next_batch_drop_last(all_indices):
```

Comment:
More tests are needed for ``next_batch``: for all combinations of batch_size, n_epochs, drop_last